client/mm/event_log.go

```diff
@@ -0,0 +1,672 @@
+ *     - quoteFees
+ *     - pl
+ *     - events
+ *       - <eventID> -> <event>
+ */
+
+var (
+	botRunsBucket        = []byte("botRuns")
+	eventsBucket         = []byte("events")
+	fiatRatesBucket      = []byte("fiatRates")
+	initialDEXBalsBucket = []byte("initialDEXBalances")
+	initialCEXBalsBucket = []byte("initialCEXBalances")
+
+	startTimeKey  = []byte("startTime")
+	endTimeKey    = []byte("endTime")
+	cfgKey        = []byte("cfg")
+	baseDeltaKey  = []byte("baseDelta")
+	quoteDeltaKey = []byte("quoteDelta")
+	baseFeesKey   = []byte("baseFees")
+	quoteFeesKey  = []byte("quoteFees")
+)
+
+func newBoltEventLogDB(ctx context.Context, path string, log dex.Logger) (*boltEventLogDB, error) {
+	db, err := bbolt.Open(path, 0600, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	err = db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(botRunsBucket)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	updateEvent := make(chan *eventUpdate, 128)
+	eventLogDB := &boltEventLogDB{
+		DB:          db,
+		log:         log,
+		updateEvent: updateEvent,
+	}
+
+	go func() {
+		eventLogDB.listenForStoreEvents(ctx)
+		db.Close()
+	}()
+
+	return eventLogDB, nil
+}
+
+// _updateEvent is called for each event that is popped off the updateEvent. If
+// the event already exists, it is updated. If it does not exist, it is added.
+// The stats for the run are also updated based on the event.
+func (db *boltEventLogDB) _updateEvent(update *eventUpdate) error {
+	return db.Update(func(tx *bbolt.Tx) error {
+		botRuns := tx.Bucket(botRunsBucket)
+		runBucket := botRuns.Bucket(update.runKey)
+		if runBucket == nil {
+			return fmt.Errorf("nil run bucket for key %x", update.runKey)
+		}
+
+		runBaseDelta := bToInt64(runBucket.Get(baseDeltaKey))
+		runQuoteDelta := bToInt64(runBucket.Get(quoteDeltaKey))
+		runBaseFees := bToUint64(runBucket.Get(baseFeesKey))
+		runQuoteFees := bToUint64(runBucket.Get(quoteFeesKey))
+
+		eventsBkt, err := runBucket.CreateBucketIfNotExists(eventsBucket)
+		if err != nil {
+			return err
+		}
+
+		eventKey := encode.Uint64Bytes(update.e.ID)
+		oldEventB := eventsBkt.Get(eventKey)
+		if oldEventB != nil {
+			ver, pushes, err := encode.DecodeBlob(oldEventB)
+			if err != nil {
+				return err
+			}
+			if ver != 0 {
+				return fmt.Errorf("unknown version %d", ver)
+			}
+			if len(pushes) != 1 {
+				return fmt.Errorf("expected 1 push for event, got %d", len(pushes))
+			}
+			var oldEvent MarketMakingEvent
+			err = json.Unmarshal(pushes[0], &oldEvent)
+			if err != nil {
+				return err
+			}
+
+			runBaseDelta -= oldEvent.BaseDelta
+			runQuoteDelta -= oldEvent.QuoteDelta
+			runBaseFees -= oldEvent.BaseFees
+			runQuoteFees -= oldEvent.QuoteFees
+		}
+
+		runBaseDelta += update.e.BaseDelta
+		runQuoteDelta += update.e.QuoteDelta
+		runBaseFees += update.e.BaseFees
+		runQuoteFees += update.e.QuoteFees
+
+		err = runBucket.Put(baseDeltaKey, int64ToB(runBaseDelta))
+		if err != nil {
+			return err
+		}
```

Comment:
The deltas are stored as part of the `MarketMakingEvent` below, but for the stats associated with these buckets, we really only care about the (difference between the) starting and ending balances, no?

Reply:
Based on the starting balances and the deltas you can determine the ending balances. By storing the deltas we can also know how much was due to the trades and how much was due to fees.

Reply:
You're overwriting with the most recent value, so much of that information is lost anyway. But my main point is that we should prefer idempotent functions. If this function was called twice with the same `update`, the bookkeeping would be off. If we instead just stored the most recent balances, it wouldn't matter.

Reply:
This is an idempotent function. If you call it twice with the same `update`, nothing will happen. If the update is already in the db, then the effect of the update being overwritten is removed from the overall run deltas, and then the deltas are updated by the values of the overwriting update. There will not be a state in which the overall run deltas are not equal to the sum of the deltas in each of the events. The reason for keeping track of these numbers is because it's more convenient than having to loop through each of the events to get the overall summary.

This is similar to how the balance tracking is done in the `unifiedExchangeAdaptor`. We don't keep track of the current balance of the bot. The current balance is the "base" balance combined with the effects of all pending events.

Reply:
Oh right. But what about a missed update?